Graphs/README.md

```diff
@@ -0,0 +1,188 @@
+	--   each root corresponds to the root of one of the trees passed to that
+	--   function.
+    roots :: r Int
+  }
+  deriving (Generic)
+
+type RootedGraph v = MultiplyRootedGraph Identity v
+```
+
+We can then take advantage of our usual `f a -> a` algebras by traversing the list of vertices, building a parallel list with the `a` in place of the `f Int`, and looking up the indices of the children as we go. If our graph is rooted, we finish by looking up the `a` for each root index.
+
+## Implementation
+
+```haskell
+-- | A graph with vertices of type `v ()` and some structure `r` of roots.
+data Graph r v
+  = Graph
+      { -- | Each entry in the list is a vertex in the graph. The `Int`s
+	    --   represent edges from the current vertex to the vertex at the index
+		--   indicated by the `Int`.
+        graphVertices :: [v Int],
+        -- | The roots of the graph are some structure (often `Identity` or
+		--  `[]`) indicating vertices that are "blessed" as some starting point
+		--   of the graph. If the graph was constructed from
+		--  `eliminateCommonSubtrees`, each root corresponds to the root of one
+		--   of the trees passed to that function.
+        graphRoots :: r Int
+      }
+  deriving (Generic)
+
+-- | Used internally to hold information we need while constructing a graph from
+--   trees.
+data GraphState f
+  = GraphState
+      { -- | These are nodes we've already come across, so we can simply return
+	    --   the existing `Int` index rather than introducing a new node in the
+		--   graph. This /could/ be satisfied by using
+		--   @`flip` `elemIndex` . `vertices`@, but that would be O(n), whereas
+		--   this is O(1).
+        seenNodes :: HMap.HashMap (f Int) Int,
+        -- | These are exactly the vertices (in reverse order) that will be in
+		--   the graph after creation.
+        vertices :: [f Int],
+        -- | Another optimization, this could be `length . vertices`, but that
+	    --   is again O(n) whereas this is O(1).
+        lastIndex :: Int
+      }
+
+-- | This builds a single graph from some structure of trees. Any common
+--   substructures in different trees will be shared the same way as common
+--   substructures within a tree.
+eliminateCommonSubtrees ::
+  ( Traversable r,
+    Recursive (->) t f,
+    Traversable f,
+    Eq (f Int),
+    Hashable (f Int)
+  ) =>
+  r t ->
+  Graph r f
+eliminateCommonSubtrees =
+  uncurry (flip Graph)
+    . fmap (reverse . vertices)
+    . flip State.runState (GraphState HMap.empty [] 0)
+    . traverse (cataM findOrAddNode)
+
+findOrAddNode ::
+  (Eq (f Int), Hashable (f Int)) =>
+  f Int ->
+  State.State (GraphState f) Int
+findOrAddNode tree = do
+  gs <- State.get
+  let seen = seenNodes gs
+  case HMap.lookup tree seen of
+    Just index -> pure index
+    Nothing -> do
+      let index = lastIndex gs
+      State.put
+        . GraphState (HMap.insert tree index seen) (tree : vertices gs)
+        $ succ index
+      pure index
+```
+
+## Benefits
```

Comment:
Ah, so some of the comparisons to trees are here! This kind of motivation should come closer to the start of a section like this. Or at least _tease_ it.